pkg/kotsclient/cluster_upgrade.go

```diff
@@ -0,0 +1,65 @@
+package kotsclient
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/replicatedhq/replicated/pkg/types"
+)
+
+type UpgradeClusterRequest struct {
+	KubernetesVersion string `json:"kubernetes_version"`
+}
+
+type UpgradeClusterResponse struct {
+	Cluster                *types.Cluster    `json:"cluster"`
+	Errors                 []string          `json:"errors"`
+	SupportedDistributions map[string]string `json:"supported_distributions"`
+}
+
+type UpgradeClusterOpts struct {
+	KubernetesVersion string
+	DryRun            bool
+}
+
+func (c *VendorV3Client) UpgradeCluster(clusterID string, opts UpgradeClusterOpts) (*types.Cluster, *ClusterValidationError, error) {
+	req := UpgradeClusterRequest{
+		KubernetesVersion: opts.KubernetesVersion,
+	}
+
+	if opts.DryRun {
+		ve, err := c.doUpgradeClusterDryRunRequest(clusterID, req)
+		return nil, ve, err
+	}
+	return c.doUpgradeClusterRequest(clusterID, req)
+}
+
+func (c *VendorV3Client) doUpgradeClusterRequest(clusterID string, req UpgradeClusterRequest) (*types.Cluster, *ClusterValidationError, error) {
+	resp := UpgradeClusterResponse{}
+	endpoint := fmt.Sprintf("/v3/cluster/%s/upgrade", clusterID)
+	err := c.DoJSON("POST", endpoint, http.StatusOK, req, &resp)
+	if err != nil {
+		if strings.Contains(err.Error(), " 400: ") {
+			// to avoid a lot of brittle string parsing, we make the request again with
+			// a dry-run flag and expect to get the same result back
+			ve, _ := c.doUpgradeClusterDryRunRequest(clusterID, req)
+			if ve != nil && len(ve.Errors) > 0 {
+				return nil, ve, nil
+			}
+		}
+		return nil, nil, err
+	}
```

Comment:
There is a potential better way, without having to assume `400:` in the error message.
https://github.com/replicatedhq/replicated/blob/main/pkg/kotsclient/release_compatibility.go#L40-L49
The `err` also contains the body response. So if it is a bad request, based on the swagger api, you could parse the returned body.

Reply:
this is copy paste from cluster_create